CIP-????/README.md

```diff
@@ -0,0 +1,97 @@
+    - Philip DiSarro <philip.disarro@iohk.io>
+    - Pi Lanningham <pi@sundaeswap.finance>
+Discussions:
+Created: 2023-12-14
+License: CC-BY-4.0
+---
+
+## Abstract
+
+Plutus Core lacks builtin data structures with good asymptotic performance for some use cases.
+
+## Problem
+
+Plutus Core has a few builtin data structures, but these are mostly used to make a minimally adequate representation of the `Data` type. 
+It does not have builtin data structures optimized for performance.
+
+Users can implement their own data structures (since Plutus Core is an expressive programming language), but in practice this has not happened much. 
+In particular, we will focus on two examples here:
+
+1. Arrays with constant-time lookup
+2. Maps with logarithmic-time lookup (also Sets, but we can treat them as a special case of Maps)
+
+Both of these are difficult to implement in Plutus Core:
+
+1. Arrays are (we believe) impossible without some kind of primitive with constant-time lookup
+2. Maps are possible but are typically moderately complex data structures which require a lot of code, and this has not been done in practice
+
+## Use cases
+
+### Arrays
+
+#### Order matching
+
+A common pattern in DEXs is to have a list of inputs/outputs to match up in a datum. 
+For example, we might have:
+```
+inputIdxs :: BuiltinList Integer
+outputIdxs :: BuiltinList Integer
+```
+We then want to go through these lists, looking up the corresponding inputs and outputs and check some property (e.g. that the value is directly transferred from one to the other).
+
+This requires a quadratic amount of work, which puts a low ceiling on how many orders can be processed at once. 
+Empirically, many are capped at about 30, whereas if they were limited only by the amount of space in the transaction for inputs and outputs the limit would be hundreds.
+
+If we had arrays with constant time indexing, we could make this linear instead (we would still need to do a linear amount of work to create arrays for the transaction inputs and outputs from the lists in the script context).
+
+#### `Constr` arguments
+
+The `Data` type has a `Constr` alternative which is used for encoding datatype constructors. 
+This is used for encoding the script context, and is used by languages such as Aiken extensively for representing user-defined datatypes also.
+
+The fields of the constructor are encoded in a list; hence to access a particular field the compiled code needs to do a linear amount of work. 
+If the arguments to a `Constr` were an array, we could access the fields in constant time.
+
+### Maps
+
+#### Operations on `Value`
+
+The `Value` type is a nested map: it is a map from bytestrings (representing policy IDs) to maps from bytestrings (representing token names) to integers (representing quantities).
+Since map operations are currently linear, this means that even simple operations like checking whether one value is less than another can have quadratic cost.
```

Comment:
> Since map operations are currently linear

To be precise, operations that should be logarithmic are currently linear (e.g., insertion, deletion, member), whereas operations should be linear (or nlogn) are currently quadratic (e.g., union)

Reply:
If we had a list with O(1) index lookup then `valueOf` would be constant time in the majority of DApps in practice. This is because every robust DApp already explicitly prevents non-protocol related tokens from being stored in protocol script UTxOs, and in the vast majority of cases the validator will have access to the currency symbol of all the protocol tokens that it must check for. In all such cases you can just predetermine the indices for the relevant protocol tokens and then use `elemAt index valuePairList` instead of `valueOf`